src/main/java/com/thebois/models/beings/roles/RoleFactory.java

```diff
@@ -0,0 +1,101 @@
+package com.thebois.models.beings.roles;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+/**
+ * Creates requested instances of specific roles.
+ */
+public final class RoleFactory {
+
+    private RoleFactory() {
+    }
+
+    /**
+     * Gets a role using its type.
+     *
+     * @param roleType The type of role to get
+     *
+     * @return The role with the specified type
+     *
+     * @throws java.util.NoSuchElementException If there is no such role
+     */
+    public static AbstractRole fromType(final RoleType roleType) {
+        return all()
+            .stream()
+            .filter(role -> role.getType().equals(roleType))
+            .findFirst()
+            .orElseThrow();
```

Comment:
Seen this at multiple places, what does it do?

Reply:
I recommend reading about the [Java Stream API](https://www.baeldung.com/java-8-streams#pipeline), it's hugely useful when working with collections of objects.

To break down what is happening here line by line:

`all()` just gets all roles.

`.stream()` just converts the `Collection<AbstractRole>` into a temporary `Stream<AbstractRole>` that can be operated on.

`.filter(role -> role.getType().equals(roleType))` does exactly what it says it does, it filters. Based on what? Based on the lambda given as the argument. In this case `role-> role.getType().equals(roleType)`. The lambda is called for every element in the collection and returns whether or not to keep the elements or exclude them.

`.findFirst()` gets the first element of the resulting filtered `Stream`, but since there might be no elements in it, it returns an `Optional<AbstractRole>` which has a method called `.orElseThrow()` which throws an error if it does not contain any value.

Reply:
So in layman's terms, it gets all roles, filters to only keep roles with a type equal to `roleType`, tries to get the first result, and throws an error if there were no roles that matched the filter.